### ISSUE TITLE
Mac client intermittently not sending "a" key

### ISSUE THREAD

**Issue by cgonzo** [Xpra-org/xpra#3026]
XPRA is occasionally getting into a state where it no longer sends the "a" key when I press it. Other keys seem to work. If I switch to another application and switch back, I can usually type a single "a" before being stuck again. Other times, it will send repeating "a"s until I switch to another application. Killing and restarting the xpra client seems to fix it temporarily. The following is on stdout when I set "-d keyboard" on the command line.
```
2021-02-19 14:06:57,582 mask_to_names(<flags 0 of type Gdk.ModifierType>)=['mod2'] swap_keys=False, modmap={<flags GDK_SHIFT_MASK of type Gdk.ModifierType>: 'shift', <flags GDK_LOCK_MASK of type Gdk.ModifierType>: 'lock', <flags GDK_SUPER_MASK of type Gdk.ModifierType>: 'mod3', <flags GDK_HYPER_MASK of type Gdk.ModifierType>: 'mod4', <flags GDK_META_MASK of type Gdk.ModifierType>: 'mod1', <flags GDK_CONTROL_MASK of type Gdk.ModifierType>: 'control'}, num_lock_state=True, num_lock_modifier=mod2
2021-02-19 14:06:57,582 parse_key_event(<Gdk.EventKey object at 0x13246a6d0 (void at 0x7f9a323698c0)>, True)=<GTKKeyEvent object, contents: {'modifiers': ['mod2'], 'keyname': 'a', 'keyval': 97, 'keycode': 0, 'group': 0, 'string': 'a', 'pressed': True}>
2021-02-19 14:06:57,582 handle_key_action(ClientWindow(179), <GTKKeyEvent object, contents: {'modifiers': ['mod2'], 'keyname': 'a', 'keyval': 97, 'keycode': 0, 'group': 0, 'string': 'a', 'pressed': True}>) wid=179
2021-02-19 14:06:57,582 key_handled_as_shortcut(ClientWindow(179), 'a', ['mod2'], True) shortcuts=None
2021-02-19 14:06:57,583 send_key_action(179, <GTKKeyEvent object, contents: {'modifiers': ['mod2'], 'keyname': 'a', 'keyval': 97, 'keycode': 0, 'group': 0, 'string': 'a', 'pressed': True}>)
2021-02-19 14:06:57,584 mask_to_names(<flags 0 of type Gdk.ModifierType>)=['mod2'] swap_keys=False, modmap={<flags GDK_SHIFT_MASK of type Gdk.ModifierType>: 'shift', <flags GDK_LOCK_MASK of type Gdk.ModifierType>: 'lock', <flags GDK_SUPER_MASK of type Gdk.ModifierType>: 'mod3', <flags GDK_HYPER_MASK of type Gdk.ModifierType>: 'mod4', <flags GDK_META_MASK of type Gdk.ModifierType>: 'mod1', <flags GDK_CONTROL_MASK of type Gdk.ModifierType>: 'control'}, num_lock_state=True, num_lock_modifier=mod2
2021-02-19 14:06:57,584 parse_key_event(<Gdk.EventKey object at 0x13246a540 (void at 0x7f9a32369c80)>, True)=<GTKKeyEvent object, contents: {'modifiers': ['mod2'], 'keyname': '0xffffff', 'keyval': 16777215, 'keycode': 0, 'group': 0, 'string': '', 'pressed': True}>
2021-02-19 14:06:57,584 handle_key_action(ClientWindow(179), <GTKKeyEvent object, contents: {'modifiers': ['mod2'], 'keyname': '0xffffff', 'keyval': 16777215, 'keycode': 0, 'group': 0, 'string': '', 'pressed': True}>) wid=179
2021-02-19 14:06:57,584 key_handled_as_shortcut(ClientWindow(179), '0xffffff', ['mod2'], True) shortcuts=None
2021-02-19 14:06:57,585 send_key_action(179, <GTKKeyEvent object, contents: {'modifiers': ['mod2'], 'keyname': '0xffffff', 'keyval': 16777215, 'keycode': 0, 'group': 0, 'string': '', 'pressed': True}>)
2021-02-19 14:06:57,684 mask_to_names(<flags 0 of type Gdk.ModifierType>)=['mod2'] swap_keys=False, modmap={<flags GDK_SHIFT_MASK of type Gdk.ModifierType>: 'shift', <flags GDK_LOCK_MASK of type Gdk.ModifierType>: 'lock', <flags GDK_SUPER_MASK of type Gdk.ModifierType>: 'mod3', <flags GDK_HYPER_MASK of type Gdk.ModifierType>: 'mod4', <flags GDK_META_MASK of type Gdk.ModifierType>: 'mod1', <flags GDK_CONTROL_MASK of type Gdk.ModifierType>: 'control'}, num_lock_state=True, num_lock_modifier=mod2
2021-02-19 14:06:57,684 parse_key_event(<Gdk.EventKey object at 0x13246a540 (void at 0x7f9a32369aa0)>, False)=<GTKKeyEvent object, contents: {'modifiers': ['mod2'], 'keyname': 'a', 'keyval': 97, 'keycode': 0, 'group': 0, 'string': 'a', 'pressed': False}>
2021-02-19 14:06:57,685 handle_key_action(ClientWindow(179), <GTKKeyEvent object, contents: {'modifiers': ['mod2'], 'keyname': 'a', 'keyval': 97, 'keycode': 0, 'group': 0, 'string': 'a', 'pressed': False}>) wid=179
2021-02-19 14:06:57,685 key_handled_as_shortcut(ClientWindow(179), 'a', ['mod2'], False) shortcuts=None
2021-02-19 14:06:57,685 send_key_action(179, <GTKKeyEvent object, contents: {'modifiers': ['mod2'], 'keyname': 'a', 'keyval': 97, 'keycode': 0, 'group': 0, 'string': 'a', 'pressed': False}>)
```

**Comment by cgonzo** [comment]
For clarification, the above prints out when 'a' is not sent to the x application. I can do another dump where it works correctly if needed.

**Comment by totaam** [comment]
@cgonzo it seems that the events are being sent to the server? (a `send_key_action` is seen for every key event?)
If that's the case, we need the server's `-d keyboard` log to understand why it doesn't actually press the corresponding key.

**Comment by SaitouToshihide** [comment]
This is the same symptom as problem #2977 and is probably something related.

https://github.com/Xpra-org/xpra/issues/2977

Just for your reference,

**Comment by cgonzo** [comment]
Here's what the server log has as I'm tapping the "a" key a couple times:

2021-02-22 15:50:15,465 will try levels: [0, 4, 1, 5, 2, 6, 3, 7]
2021-02-22 15:50:15,466 do_get_keycode(0, 'a', True, ['mod2'], 0)=38 (level=0, shift=False, mode=0) 
2021-02-22 15:50:15,466 process_key_action(['key-action', 171, 'a', True, ('mod2',), 97, 'a', 0, 0]) server keycode=38, group=0 
2021-02-22 15:50:15,466 set_keyboard_layout_group(0) config=KeyboardConfig(us /  / ), current keyboard group=0 
2021-02-22 15:50:15,466 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,466 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,467 is_modifier(38) not found
2021-02-22 15:50:15,467 handle_key((171, True, 'a', 97, 38, ['mod2'], False, True))
2021-02-22 15:50:15,467 handle keycode 38: key a was already pressed, ignoring
2021-02-22 15:50:15,467 process_key_action(['key-action', 171, '0xffffff', True, ('mod2',), 16777215, '', 0, 0]) server keycode=-1, group=0 
2021-02-22 15:50:15,467 set_keyboard_layout_group(0) config=KeyboardConfig(us /  / ), current keyboard group=0 
2021-02-22 15:50:15,467 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,467 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,563 process_key_action(['key-action', 171, 'a', False, ('mod2',), 97, 'a', 0, 0]) server keycode=-1, group=0 
2021-02-22 15:50:15,563 set_keyboard_layout_group(0) config=KeyboardConfig(us /  / ), current keyboard group=0 
2021-02-22 15:50:15,563 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,563 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,722 will try levels: [0, 4, 1, 5, 2, 6, 3, 7]
2021-02-22 15:50:15,722 do_get_keycode(0, 'a', True, ['mod2'], 0)=38 (level=0, shift=False, mode=0) 
2021-02-22 15:50:15,723 process_key_action(['key-action', 171, 'a', True, ('mod2',), 97, 'a', 0, 0]) server keycode=38, group=0 
2021-02-22 15:50:15,723 set_keyboard_layout_group(0) config=KeyboardConfig(us /  / ), current keyboard group=0 
2021-02-22 15:50:15,723 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,723 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,723 is_modifier(38) not found
2021-02-22 15:50:15,723 handle_key((171, True, 'a', 97, 38, ['mod2'], False, True))
2021-02-22 15:50:15,724 handle keycode 38: key a was already pressed, ignoring
2021-02-22 15:50:15,724 process_key_action(['key-action', 171, '0xffffff', True, ('mod2',), 16777215, '', 0, 0]) server keycode=-1, group=0 
2021-02-22 15:50:15,724 set_keyboard_layout_group(0) config=KeyboardConfig(us /  / ), current keyboard group=0 
2021-02-22 15:50:15,724 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,724 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,789 process_key_action(['key-action', 171, 'a', False, ('mod2',), 97, 'a', 0, 0]) server keycode=-1, group=0 
2021-02-22 15:50:15,790 set_keyboard_layout_group(0) config=KeyboardConfig(us /  / ), current keyboard group=0 
2021-02-22 15:50:15,790 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,790 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,964 will try levels: [0, 4, 1, 5, 2, 6, 3, 7]
2021-02-22 15:50:15,964 do_get_keycode(0, 'a', True, ['mod2'], 0)=38 (level=0, shift=False, mode=0)
2021-02-22 15:50:15,964 process_key_action(['key-action', 171, 'a', True, ('mod2',), 97, 'a', 0, 0]) server keycode=38, group=0
2021-02-22 15:50:15,964 set_keyboard_layout_group(0) config=KeyboardConfig(us /  / ), current keyboard group=0
2021-02-22 15:50:15,965 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,965 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,965 is_modifier(38) not found
2021-02-22 15:50:15,965 handle_key((171, True, 'a', 97, 38, ['mod2'], False, True))
2021-02-22 15:50:15,965 handle keycode 38: key a was already pressed, ignoring
2021-02-22 15:50:15,965 process_key_action(['key-action', 171, '0xffffff', True, ('mod2',), 16777215, '', 0, 0]) server keycode=-1, group=0
2021-02-22 15:50:15,965 set_keyboard_layout_group(0) config=KeyboardConfig(us /  / ), current keyboard group=0
2021-02-22 15:50:15,966 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:15,966 filtered_modifiers_set(['mod2'])=set(['mod2'])
2021-02-22 15:50:16,072 process_key_action(['key-action', 171, 'a', False, ('mod2',), 97, 'a', 0, 0]) server keycode=-1, group=0
2021-02-22 15:50:16,072 set_keyboard_layout_group(0) config=KeyboardConfig(us /  / ), current keyboard group=0

**Comment by totaam** [comment]
Thanks for the logs. I think we've just uncovered a huge keyboard bug.

TIL: `handle keycode 38: key a was already pressed, ignoring`
That's because, for the key press we get
```
do_get_keycode(0, 'a', True, ['mod2'], 0)=38 (level=0, shift=False, mode=0) 
process_key_action(['key-action', 171, 'a', True, ('mod2',), 97, 'a', 0, 0]) server keycode=38, group=0
```
But for the key release:
```
process_key_action(['key-action', 171, 'a', False, ('mod2',), 97, 'a', 0, 0]) server keycode=-1, group=0
```

We always try to unpress the same keycode that we pressed before to ensure that keys don't get wedged.
Unfortunately, this code was buggy and stored `-1` in the key translation map when it was meant to ignore it..
That's a 2 year old bug introduced in 7f348be7be0051da950c89c92bc85455e5649669.

Fixed in 0408312ea01318cd113a7f113577a25fb62a4de6, which will need to be backported.
In the meantime, you can apply it by hand as it is very small (one small line changed).

---

Also, looking again at you client log, in between the key press and key release, there is a:
```
handle_key_action(ClientWindow(179), <GTKKeyEvent object, contents: {'modifiers': ['mod2'], 'keyname': '0xffffff', 'keyval': 16777215, 'keycode': 0, 'group': 0, 'string': '', 'pressed': True}>) wid=179
```
That's an empty event, which we ignore. But perhaps it has a meaning.
